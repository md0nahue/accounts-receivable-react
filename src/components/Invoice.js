import React from "react";
import { connect } from "react-redux";

const mapStateToProps = (state, { match }) => {
  return {
    invoice: state.invoices.find(invoice => invoice.id === match.params.id)
  };
};

function Invoice(props) {
  if (props.invoice) {
    return (
      <div className="Invoice">
        <h2>{props.invoice.client} Invoice</h2>
        <div>
          <strong>Attn: </strong>
          {props.invoice.attn}
        </div>
        <div>
          <strong>Due Date: </strong>
          {props.invoice.due_date}
        </div>
        <div>
          <strong>Notes: </strong>
          {props.invoice.notes}
        </div>
        <br />
        <h5>Products</h5>
        <table class="table">
          <thead>
            <tr>
              <th scope="col">Name</th>
              <th scope="col">Description</th>
              <th scope="col">Unit Price</th>
              <th scope="col">Quantity</th>
              <th scope="col">Subtotal</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>Anti-virus Software</td>
              <td>Anti-virus software subscription</td>
              <td>$200</td>
              <td>1</td>
              <td>$200</td>
            </tr>
          </tbody>
        </table>
        <br />
        <h5>Services</h5>
        <table class="table">
          <thead>
            <tr>
              <th scope="col">Name</th>
              <th scope="col">Description</th>
              <th scope="col">Unit Price</th>
              <th scope="col">Quantity</th>
              <th scope="col">Subtotal</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>Anti-virus Software</td>
              <td>Anti-virus software subscription</td>
              <td>$200</td>
              <td>1</td>
              <td>$200</td>
            </tr>
          </tbody>
        </table>
        <div>
          <strong>Total: </strong>
          $1,000,000
        </div>
      </div>
    );
  } else {
    return (
      <div className="Invoice">
        <h3>Select an invoice</h3>
      </div>
    );
  }
}

export default connect(mapStateToProps)(Invoice);
