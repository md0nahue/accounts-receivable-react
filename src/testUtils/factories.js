export const lineItemFactory = (overrides = {}) => {
  return {
    name: "Anti-virus Software",
    description: "Subscription service for virus software.",
    discount_percent: null,
    price_cents: 20000,
    quantity: 1,
    ...overrides,
  };
};

let nextInvoiceId = 1;

export const invoiceFactory = (overrides = {}) => {
  return {
    id: String(nextInvoiceId++),
    client: "CircleCi",
    attn: "Bo <bo@circleci.com>",
    due_date: "10/10/2020",
    notes: "Send monthly invoices.",
    projects: [lineItemFactory()],
    services: [
      lineItemFactory({
        name: "Software Development",
        description: "Deploy new environment.",
        price_cents: 15000,
      }),
    ],
    ...overrides,
  };
};
