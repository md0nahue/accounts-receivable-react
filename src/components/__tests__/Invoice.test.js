import React from "react";
import { render, screen } from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { MemoryRouter, Route } from "react-router-dom";

import { invoiceFactory } from "../../testUtils/factories";
import { Invoice } from "../Invoice";

jest.mock("../InvoiceSection", () => ({
  InvoiceSection: (props) => `InvoiceSection:${JSON.stringify(props)}`,
}));

const mockStore = configureStore();

describe("Invoice", () => {
  it("Renders an invoice matching url 'id' param", () => {
    const invoices = [
      invoiceFactory({ id: "1", client: "Microsoft" }),
      invoiceFactory({ id: "2", client: "Netflix" }),
      invoiceFactory({ id: "3", client: "Facebook" }),
    ];

    const store = mockStore({ invoices });

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/invoices/2"]}>
          <Route path="/invoices/:id">
            <Invoice />
          </Route>
        </MemoryRouter>
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders invoice total", () => {
    const invoices = [invoiceFactory({ id: "1", client: "Microsoft" })];

    const store = mockStore({ invoices });

    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/1"]}>
          <Route path="/:id">
            <Invoice />
          </Route>
        </MemoryRouter>
      </Provider>,
    );

    const totalDiv = screen.queryByTestId(`invoice-1-total`);

    expect(totalDiv).toBeTruthy();
    expect(totalDiv.textContent).toEqual("Total: $350");
  });
});
