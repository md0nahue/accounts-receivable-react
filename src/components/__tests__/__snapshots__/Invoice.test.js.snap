// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice Renders an invoice matching url 'id' param 1`] = `
<DocumentFragment>
  <div
    class="Invoice"
  >
    <h2>
      Netflix Invoice
    </h2>
    <div>
      <strong>
        Attn: 
      </strong>
      Bo &lt;bo@circleci.com&gt;
    </div>
    <div>
      <strong>
        Due Date: 
      </strong>
      10/10/2020
    </div>
    <div>
      <strong>
        Notes: 
      </strong>
      Send monthly invoices.
    </div>
    <br />
    InvoiceSection:{"title":"Projects","items":[{"name":"Anti-virus Software","description":"Subscription service for virus software.","discount_percent":null,"price_cents":20000,"quantity":1}]}
    <br />
    InvoiceSection:{"title":"Services","items":[{"name":"Software Development","description":"Deploy new environment.","discount_percent":null,"price_cents":15000,"quantity":1}]}
    <div
      data-testid="invoice-2-total"
    >
      <strong>
        Total: 
      </strong>
      $1,000,000
    </div>
    <div>
      <button
        class="button complete"
      >
        Mark as complete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
